zebra-state/src/service/check/tests/prop.rs

```diff
@@ -0,0 +1,386 @@
+//! Randomised property tests for state contextual validation nullifier: (), in_finalized_state: ()  nullifier: (), in_finalized_state: ()  checks.
+
+use std::sync::Arc;
+
+use itertools::Itertools;
+use proptest::prelude::*;
+
+use zebra_chain::{
+    block::{Block, Height},
+    fmt::TypeNameToDebug,
+    parameters::Network::*,
+    primitives::Groth16Proof,
+    serialization::ZcashDeserializeInto,
+    sprout::{self, JoinSplit},
+    transaction::{JoinSplitData, LockTime, Transaction},
+};
+
+use crate::{
+    config::Config, service::StateService, tests::Prepare, FinalizedBlock,
+    ValidateContextError::DuplicateSproutNullifier,
+};
+
+// These tests use the `Arbitrary` trait to easily generate complex types,
+// then modify those types to cause an error (or to ensure success).
+//
+// We could use mainnet or testnet blocks in these tests,
+// but the differences shouldn't matter,
+// because we're only interested in spend validation,
+// (and passing various other state checks).
+proptest! {
+    /// Make sure an arbitrary sprout nullifier is accepted by state contextual validation.
+    ///
+    /// This test makes sure there are no spurious rejections that might hide bugs in the other tests.
+    /// (And that the test infrastructure generally works.)
+    #[test]
+    fn accept_distinct_arbitrary_sprout_nullifiers(
+        mut joinsplit in TypeNameToDebug::<JoinSplit::<Groth16Proof>>::arbitrary(),
+        mut joinsplit_data in TypeNameToDebug::<JoinSplitData::<Groth16Proof>>::arbitrary(),
+    ) {
```

Comment:
TODO: randomly commit the final block to the finalized or non-finalized state.

This change can happen in a future PR.

Reply:
Fixed in dad6ce51